experimental/dds/tree2/src/test/shared-tree/editing.spec.ts

```diff
@@ -1012,6 +1038,199 @@ describe("Editing", () => {
+				tree.editor.addNodeExistsConstraint({
+					parent: rootPath,
+					parentField: brand("foo"),
+					parentIndex: 0,
+				});
+				const rootSequence = tree.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				rootSequence.insert(1, singleTextCursor({ type: jsonObject.name }));
+				tree.transaction.commit();
+
+				// Make a concurrent edit to rebase over that inserts into root sequence
+				// State should be (to tree2): [{}, {}]
+				const tree2RootSequence = tree2.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				tree2RootSequence.insert(1, singleTextCursor({ type: jsonObject.name }));
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], [{ foo: {} }, {}, {}]);
+			});
+
+			it.skip("constraint on newly inserted node with ancestor deleted concurrently", () => {
+				const tree = makeTreeFromJson([{}]);
+				const tree2 = tree.fork();
+
+				const rootPath: UpPath = {
+					parent: undefined,
+					parentField: rootFieldKeySymbol,
+					parentIndex: 0,
+				};
+
+				// Insert "a"
+				// State should be: [{ foo: "a"}]
+				const sequence = tree.editor.sequenceField({
+					parent: rootPath,
+					field: brand("foo"),
+				});
+				sequence.insert(0, singleTextCursor({ type: jsonString.name, value: "a" }));
+
+				// Constrain on "a" existing and insert "b" into root sequence
+				// State should be (if "a" exists): [{ foo: "a"}, "b"]
+				// a's ancestor will be concurrently deleted
+				tree.transaction.start();
+				tree.editor.addNodeExistsConstraint({
+					parent: rootPath,
+					parentField: brand("foo"),
+					parentIndex: 0,
+				});
+				const rootSequence = tree.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				rootSequence.insert(1, singleTextCursor({ type: jsonString.name, value: "b" }));
+				tree.transaction.commit();
+
+				// Concurrently delete a's ancestor
+				// State should be (to tree2): []
+				const tree2RootSequence = tree2.editor.sequenceField({
+					parent: undefined,
+					field: rootFieldKeySymbol,
+				});
+				tree2RootSequence.delete(0, 1);
+
+				tree.merge(tree2);
+				tree2.rebaseOnto(tree);
+
+				expectJsonTree([tree, tree2], []);
+			});
+
+			it("constraint works with node created as part of transaction", () => {
+				const tree = makeTreeFromJson([{}]);
+				const tree2 = tree.fork();
+
+				const rootPath: UpPath = {
+					parent: undefined,
+					parentField: rootFieldKeySymbol,
+					parentIndex: 0,
+				};
+
+				// Insert "a" under foo
+				// State should be: [{ foo: "a" }]
+				const sequence = tree.editor.sequenceField({
+					parent: rootPath,
+					field: brand("foo"),
+				});
+				sequence.insert(0, singleTextCursor({ type: jsonString.name, value: "a" }));
+
+				// Constrain on "a" existing and insert "b" if it does
+				// State should be (if "a" exists): [{ foo: "a"}, "b"]
+				tree.transaction.start();
```

Comment:
I think the transaction is supposed to include the insert of "a".

Reply:
Moved "a" insert to be inside the transaction.